samples/solutions/stingy_connection.py

```diff
@@ -0,0 +1,275 @@
+            randint = random.randint(UPPER_LIMIT_OF_ENQUEUEING, UPPER_LIMIT_OF_ENQUEUEING)
+            self.log_info_and_print(
+                "Will sleep for {} seconds and then enqueue messages".format(randint)
+            )
+            await asyncio.sleep(randint)
+            message_id += 1
+
+    async def do_all_tasks_and_disconnect(self):
+        self.log_info_and_print("Current qsize is: {}".format(self.message_queue.qsize()))
+        while not self.message_queue.empty():
+            msg = await self.message_queue.get()
+            task = asyncio.create_task(self.device_client.send_message(msg))
+            self.task_list.append(task)
+            self.message_queue.task_done()
+        try:
+            if self.device_client.connected:
+                self.log_info_and_print("sending messages...")
+                await asyncio.gather(*self.task_list)
+                self.log_info_and_print("sent all messages...")
+                self.task_list = []
+                await self.device_client.disconnect()
+        except Exception as e:
+            self.log_error_and_print(
+                "Caught exception while trying to send message or disconnect: {}".format(
+                    get_type_name(e)
+                )
+            )
+
+    async def connect_with_retry_send_all_and_disconnect(self):
+        while True:
+            for i in range(1, 10):
+                if not self.device_client.connected:
+                    try:
+                        self.log_info_and_print(
+                            "Attempting to connect the device client try number {}....".format(i)
+                        )
+                        await self.device_client.connect()
+                        self.log_info_and_print("Successfully connected the device client...")
+                        await self.do_all_tasks_and_disconnect()
+                        break
+                    except Exception as e:
+                        self.log_error_and_print("Caught exception while trying to connect...")
+                        if type(e) is exceptions.CredentialError:
+                            self.log_error_and_print(
+                                "Failed to connect the device client due to incorrect or badly formatted credentials..."
+                            )
+                            raise
+                        if self.is_retryable(e):
+                            self.log_error_and_print(
+                                "Failed to connect the device client due to retryable error."
+                                "Sleeping and retrying after some time..."
+                            )
+                            await asyncio.sleep(CONNECTION_ATTEMPT_INTERVAL)
+                        else:
+                            self.log_error_and_print(
+                                "Failed to connect the device client due to not-retryable error...."
+                            )
+                            raise
```

Comment:
>raise [](http://example.com/codeflow?start=28&length=5)

This task fails as soon as there is a non-retryable error. Is this what we want? I don't think customers would want the app to exit in this case.

Reply:
i think i would have to discuss for this coz this is pretty much the same behavior i did for the other sample. If the error is not retry-able the customer probably has to change some configurational thing which could be credentials or something else. In that scenario I thought the best case is to exit the app and start again. In case our error situation become more fine tuned then we can reduce this list of non-retryable errors.